File: npm/llmrouter/src/serve.ts
/**
 * `llmrouter serve` — the decision engine as an HTTP decide-sidecar.
 *
 *   POST /decide   {RoutingRequest payload}  -> {Decision payload}
 *   GET  /stats    session/backend telemetry
 *   GET  /healthz  liveness
 *
 * Wire-compatible with the Python `llmrouter serve` (same snake_case payloads)
 * and with `llmrouter-state-v1` snapshots via store.ts, so a LiteLLM/Bifrost/
 * custom gateway can point at either runtime.
 *
 * Feedback loop across the HTTP boundary: include
 *   "observed_usage": {cache_read_tokens, cache_write_tokens, ...}
 * with the NEXT request — the counters the gateway saw from the provider —
 * and the learned hit rate for this session is updated before deciding.
 */
import { createServer, type IncomingMessage, type Server, type ServerResponse } from "node:http";
import { readFileSync, renameSync, writeFileSync } from "node:fs";
import { PolicyEngine, DEFAULT_CONFIG, type RouterConfig } from "./policy.js";
import { BackendRegistry, BackendSpec } from "./backends.js";
import { QualityModel } from "./learning.js";
import { PriceRegistry } from "./pricing.js";
import { WorkloadResolver } from "./signals.js";
import { restoreSnapshot, takeSnapshot, type Snapshot } from "./store.js";
import { DeploymentClass, RoutingRequest } from "./types.js";

const MAX_BODY_BYTES = 10 * 1024 * 1024;

export interface Sidecar {
  server: Server;
  engine: PolicyEngine;
  statePath: string | null;
}

export function buildEngine(
  qualityModel: QualityModel | null = null, config?: RouterConfig,
): PolicyEngine {
  const prices = new PriceRegistry();
  const ids = prices.ids();
  const backends = new BackendRegistry([
    new BackendSpec("anthropic", {
      deployment: DeploymentClass.Api,
      models: ids.filter((m) => m.startsWith("claude")),
      priority: 0,
    }),
    new BackendSpec("openai", {
      deployment: DeploymentClass.Api,
      models: ids.filter((m) => m.startsWith("gpt")),
      priority: 1,
    }),
  ]);
  return new PolicyEngine(
    prices, backends, new WorkloadResolver(), config ?? DEFAULT_CONFIG,
    null, qualityModel,
  );
}

function readBody(req: IncomingMessage): Promise<string> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    let size = 0;
    req.on("data", (c: Buffer) => {
      size += c.length;
      if (size > MAX_BODY_BYTES) {
        reject(new Error("body too large"));
        req.destroy();
        return;
      }
      chunks.push(c);
    });
    req.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
    req.on("error", reject);
  });
}

function send(res: ServerResponse, status: number, obj: unknown): void {
  const body = JSON.stringify(obj);
  res.writeHead(status, { "content-type": "application/json" });
  res.end(body);
}

export function createDecideServer(options: {
  host?: string;
  port?: number;
  token?: string | null;
  qualityModelPath?: string | null;
  statePath?: string | null;
  engine?: PolicyEngine;
} = {}): Sidecar {
  const qualityModel = options.qualityModelPath
    ? QualityModel.fromJson(JSON.parse(readFileSync(options.qualityModelPath, "utf8")))
    : null;
  const engine = options.engine ?? buildEngine(qualityModel);

  let sidecar: Sidecar; // declared before the handler closes over it
  const handler = async (req: IncomingMessage, res: ServerResponse) => {
    if (options.token) {
      const got = String(req.headers.authorization ?? "");
      if (got !== `Bearer ${options.token}`) {
        return send(res, 401, { error: "unauthorized" });
      }
    }
    const path = (req.url ?? "").split("?")[0];
    if (req.method === "GET" && path === "/healthz") {
      return send(res, 200, { ok: true, engine: "llmrouter" });
    }
    if (req.method === "GET" && path === "/stats") {
      const backends: Record<string, unknown> = {};
      for (const h of engine.backends.allHealth()) {
        backends[h.backendId] = {
          in_flight: h.inFlight, error_rate: h.errorRate,
          ewma_latency_ms: h.ewmaLatencyMs,
          consecutive_failures: h.consecutiveFailures,
          circuit_open: h.openUntil ? 1 : 0,
        };
      }
      return send(res, 200, {
        sessions: engine.sessions.stats(), backends,
        fingerprints: engine.resolver.fingerprints.cardinalityByClient(),
      });
    }
    if (req.method === "POST" && path === "/decide") {
      let payload: Record<string, any>;
      try {
        payload = JSON.parse(await readBody(req));
        if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
          throw new Error("body must be a JSON object");
        }
      } catch (err) {
        return send(res, 400, {
          error: `invalid JSON: ${err instanceof Error ? err.message : err}`,
        });
      }
      let decided: RoutingRequest;
      try {
        decided = RoutingRequest.fromPayload(payload);
      } catch (err) {
        return send(res, 400, { error: err instanceof Error ? err.message : err });
      }
      // Fold the previous turn's provider counters in FIRST, so this decision
      // already sees the learned hit rate (same order as the Python sidecar).
      const usage = payload.observed_usage;
      if (usage) {
        const sess = engine.sessions.peek(decided.stickyKey());
        if (sess) {
          sess.recordUsage(
            Number(usage.cache_read_tokens ?? 0),
            Number(usage.cache_write_tokens ?? 0),
          );
          engine.sessions.put(sess);
        }
      }
      try {
        const decision = engine.decide(decided);
        engine.recordDecision(decided, decision, usage ? {
          cacheReadTokens: Number(usage.cache_read_tokens ?? 0),
          cacheWriteTokens: Number(usage.cache_write_tokens ?? 0),
        } : undefined);
        if (sidecar?.statePath) {
          saveSnapshotFile(sidecar.statePath, sidecar.engine);
        }
        return send(res, 200, decision.toPayload());
      } catch (err) {
        return send(res, 422, {
          error: err instanceof Error ? err.message : String(err),
        });
      }
    }
    return send(res, 404, { error: "not found" });
  };

  const server = createServer(handler);
  sidecar = {
    server, engine,
    statePath: options.statePath ?? null,
  };
  if (options.statePath) {
    loadSnapshotFile(options.statePath, engine);
  }
  return sidecar;
}

// --------------------------------------------------------------------------- //
// Durable state (atomic: tmp + rename)
// --------------------------------------------------------------------------- //
export function saveSnapshotFile(path: string, engine: PolicyEngine): void {
  const tmp = `${path}.tmp`;
  writeFileSync(tmp, JSON.stringify(takeSnapshot(engine), null, 1));
  renameSync(tmp, path);
}

export function loadSnapshotFile(path: string, engine: PolicyEngine): boolean {
  try {
    const snap = JSON.parse(readFileSync(path, "utf8")) as Snapshot;
    restoreSnapshot(engine, snap);
    return true;
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") return false; // first boot
    throw err;
  }
}

// --------------------------------------------------------------------------- //
export function serveMain(argv: string[]): number {
  const get = (flag: string): string | undefined => {
    const i = argv.indexOf(`--${flag}`);
    return i >= 0 && i + 1 < argv.length ? argv[i + 1] : undefined;
  };
  const host = get("host") ?? "127.0.0.1";
  const port = Number(get("port") ?? 8787);
  const sc = createDecideServer({
    token: get("token") ?? null,
    qualityModelPath: get("quality-model") ?? null,
    statePath: get("state") ?? null,
  });
  sc.server.listen(port, host, () => {
    console.log(
      `llmrouter decide-sidecar on http://${host}:${port}  ` +
        `(POST /decide, GET /stats, GET /healthz)` +
        (get("token") ? "  [auth: bearer token required]" : ""),
    );
  });
  const stop = () => {
    if (sc.statePath) saveSnapshotFile(sc.statePath, sc.engine);
    sc.server.close();
    process.exit(0);
  };
  process.on("SIGINT", stop);
  process.on("SIGTERM", stop);
  return 0;
}

// Direct execution: node dist/serve.js --port 8787
if (process.argv[1] && process.argv[1].endsWith("serve.js")) {
  serveMain(process.argv.slice(2));
}

File: llmrouter/src/llmrouter/serve.py
"""`llmrouter serve` — the decision engine as an HTTP sidecar.

Zero dependencies (stdlib `http.server`), single-purpose: a gateway that already
owns its transport (LiteLLM custom router, Bifrost routing rule, Envoy ext_authz,
your own proxy) asks *us* which (model, backend) to use and executes the call
itself. This is the integration surface that lets the engine ride alongside
production gateways instead of replacing them.

    POST /decide   {RoutingRequest payload}          -> {Decision payload}
    GET  /stats    routing/session/backend telemetry -> {...}
    GET  /healthz  liveness for load balancers       -> {"ok": true}

Feedback closes the loop here too: when the caller's *next* request includes
`"observed_usage": {"cache_read_tokens": N, "cache_write_tokens": M}` (the
provider counters from the previous turn), the sidecar folds them into session
affinity — the learned-hit-rate machinery works across the HTTP boundary.

Security posture: binds 127.0.0.1 by default; pass --token to require
`Authorization: Bearer <token>`. This is a sidecar, not a public endpoint.

The Decision/Request JSON mirrors the TypeScript `serve.ts` byte-for-byte, so
one gateway client works against either runtime.
"""
from __future__ import annotations

import argparse
import json
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any, Mapping

from .learning import QualityModel
from .router import Router
from .types import RoutingRequest

MAX_BODY_BYTES = 10 * 1024 * 1024


class _Handler(BaseHTTPRequestHandler):
    server: "_Sidecar"

    # ---- plumbing --------------------------------------------------------- #
    def log_message(self, fmt: str, *args: Any) -> None:  # quiet by default
        if self.server.verbose:
            super().log_message(fmt, *args)

    def _send(self, status: int, obj: Mapping[str, Any]) -> None:
        body = json.dumps(obj).encode("utf-8")
        self.send_response(status)
        self.send_header("content-type", "application/json")
        self.send_header("content-length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def _authorized(self) -> bool:
        token = self.server.token
        if token is None:
            return True
        got = self.headers.get("authorization", "")
        return got == f"Bearer {token}"

    # ---- routes ----------------------------------------------------------- #
    def do_GET(self) -> None:  # noqa: N802 — BaseHTTPRequestHandler API
        if not self._authorized():
            return self._send(401, {"error": "unauthorized"})
        if self.path == "/healthz":
            return self._send(200, {"ok": True, "engine": "llmrouter"})
        if self.path == "/stats":
            return self._send(200, self.server.router.stats())
        return self._send(404, {"error": "not found"})

    def do_POST(self) -> None:  # noqa: N802
        if not self._authorized():
            return self._send(401, {"error": "unauthorized"})
        if self.path != "/decide":
            return self._send(404, {"error": "not found"})
        try:
            length = int(self.headers.get("content-length") or 0)
            if length <= 0 or length > MAX_BODY_BYTES:
                return self._send(400, {"error": "body must be 10MB or less"})
            payload = json.loads(self.rfile.read(length))
            if not isinstance(payload, dict):
                return self._send(400, {"error": "body must be a JSON object"})
        except (ValueError, json.JSONDecodeError) as exc:
            return self._send(400, {"error": f"invalid JSON: {exc}"})

        try:
            req = RoutingRequest.from_payload(payload)
        except (ValueError, KeyError) as exc:
            return self._send(400, {"error": str(exc)})

        try:
            with self.server.lock:
                # Feed the previous turn's provider counters in first, so THIS
                # decision already sees the learned hit rate.
                usage = payload.get("observed_usage")
                if usage:
                    from .adapters import Usage
                    self.server.router.note_usage(
                        req,
                        Usage(
                            input_tokens=int(usage.get("input_tokens", 0) or 0),
                            output_tokens=int(usage.get("output_tokens", 0) or 0),
                            cache_read_tokens=int(usage.get("cache_read_tokens", 0) or 0),
                            cache_write_tokens=int(usage.get("cache_write_tokens", 0) or 0),
                        ),
                    )
                decision = self.server.router.decide(req)
                # Pin the session for THIS turn so the next one is sticky —
                # the caller executes; we keep the affinity memory.
                self.server.router.note_decision(req, decision)
        except RuntimeError as exc:  # every candidate filtered out
            return self._send(422, {"error": str(exc)})
        return self._send(200, decision.to_payload())


class _Sidecar(ThreadingHTTPServer):
    daemon_threads = True

    def __init__(self, router: Router, host: str, port: int,
                 token: str | None, verbose: bool) -> None:
        super().__init__((host, port), _Handler)
        self.router = router
        self.token = token
        self.verbose = verbose
        # SessionStore and BackendHealth are plain dicts — one lock keeps the
        # decide+learn critical section honest under ThreadingHTTPServer.
        self.lock = threading.Lock()


def build_router(quality_model_path: str | None = None,
                 state_path: str | None = None) -> Router:
    """Decision-capable router over the bundled defaults (no keys, no network)."""
    quality_model = (QualityModel.load(quality_model_path)
                     if quality_model_path else None)
    state_store = None
    if state_path:
        from .store import FileStateStore
        # autosave every decision: sidecar traffic is low-rate and durable
        # affinity is the entire point of running one.
        state_store = FileStateStore(state_path, autosave_every=1)
    return Router.pure(quality_model=quality_model, state_store=state_store)


def create_server(router: Router, host: str = "127.0.0.1", port: int = 8787,
                  token: str | None = None, verbose: bool = False) -> _Sidecar:
    return _Sidecar(router, host, port, token, verbose)


def cmd_serve(args: argparse.Namespace) -> int:
    router = build_router(args.quality_model, args.state)
    httpd = create_server(router, args.host, args.port, args.token, args.verbose)
    where = f"http://{args.host}:{args.port}"
    print(f"llmrouter decide-sidecar on {where}  "
          f"(POST {where}/decide, GET {where}/stats, GET {where}/healthz)"
          + ("  [auth: bearer token required]" if args.token else ""))
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print("\nshutting down")
    finally:
        if args.state:
            router.save_state()
        httpd.server_close()
    return 0


def main(argv: list[str] | None = None) -> int:
    """Standalone entry (`python -m llmrouter.serve`); normally reached via
    `python -m llmrouter serve` in cli.main."""
    parser = argparse.ArgumentParser(prog="llmrouter serve")
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=8787)
    parser.add_argument("--token", default=None)
    parser.add_argument("--quality-model", default=None)
    parser.add_argument("--state", default=None)
    parser.add_argument("--verbose", action="store_true")
    args = parser.parse_args(argv)
    return cmd_serve(args)


__all__ = ["build_router", "create_server", "cmd_serve", "MAX_BODY_BYTES"]
